### PR TITLE
Fix thunks in Byron and CertFakeVRF

### PR DESCRIPTION
**Pull request by mrBliss**
With the fixes of this PR, https://github.com/input-output-hk/cardano-base/pull/126, and some local fixes in consensus, I don't get any more unexpected thunks when running the consensus tests.

### CODE REVIEW

File: byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs [hunk 1]
```diff
@@ -298,10 +357,8 @@ registerProposalComponents env rs proposal = do
   SoftwareVersion appName appVersion = softwareVersion
 
   softwareVersionChanged =
-    maybe True ((/= appVersion) . fst3) $ M.lookup appName appVersions
-    where
-      fst3 :: (a, b, c) -> a
-      fst3 (x, _, _) = x
```
Comment (nc6) [inline byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs:304]: :+1:

File: byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs [hunk 2]
```diff
@@ -99,8 +123,43 @@ data State = State
...
+    deriving anyclass (NFData, NoUnexpectedThunks)
+
+instance FromCBOR ProtocolUpdateProposal where
+  fromCBOR = do
+    enforceSize "ProtocolUpdateProposal" 2
+    ProtocolUpdateProposal <$> fromCBOR <*> fromCBOR
+
+instance ToCBOR ProtocolUpdateProposal where
+    toCBOR pup =
+      encodeListLen 2
```
Comment (nc6) [inline byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs:139]: This matches the encoding for a 2-tuple :+1:

File: byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs [hunk 3]
```diff
@@ -88,7 +93,26 @@ data Environment = Environment
...
+    deriving anyclass (NFData, NoUnexpectedThunks)
+
+instance FromCBOR ApplicationVersion where
+  fromCBOR = do
+    enforceSize "ApplicationVersion" 3
+    ApplicationVersion <$> fromCBOR <*> fromCBOR <*> fromCBOR
+
+instance ToCBOR ApplicationVersion where
+    toCBOR av =
+      encodeListLen 3
```
Comment (nc6) [inline byron/ledger/impl/src/Cardano/Chain/Update/Validation/Registration.hs:110]: Matches the instance for a 3-tuple

**Review by nc6 — APPROVED** [review]
This looks good to me. Serialisation tests all seem to still pass, so I'm happy!

**Comment by mrBliss** [comment]
Hmm, CI failed because `cardano-ledger-test` timed out on darwin. But it didn't time out on linux...

**Comment by nc6** [comment]
> Hmm, CI failed because `cardano-ledger-test` timed out on darwin. But it didn't time out on linux...

I'll restart it

**Pull request merged**